File: engine/src/solver/modal.rs
use crate::types::*;
use crate::linalg::*;
use std::collections::HashMap;
use super::dof::DofNumbering;
use super::assembly::*;
use super::mass_matrix::*;

/// Modal analysis result.
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModalResult {
    pub modes: Vec<ModeShape>,
    pub n_dof: usize,
    pub total_mass: f64,
    pub cumulative_mass_ratio_x: f64,
    pub cumulative_mass_ratio_y: f64,
    pub rayleigh: Option<RayleighDamping>,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModeShape {
    pub frequency: f64,
    pub period: f64,
    pub omega: f64,
    pub displacements: Vec<Displacement>,
    pub participation_x: f64,
    pub participation_y: f64,
    pub effective_mass_x: f64,
    pub effective_mass_y: f64,
    pub mass_ratio_x: f64,
    pub mass_ratio_y: f64,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RayleighDamping {
    pub a0: f64,
    pub a1: f64,
    pub omega1: f64,
    pub omega2: f64,
    pub damping_ratios: Vec<f64>,
}

/// Solve 2D modal analysis.
/// Solves K·φ = ω²·M·φ (generalized eigenvalue problem).
pub fn solve_modal_2d(
    input: &SolverInput,
    densities: &HashMap<String, f64>,
    num_modes: usize,
) -> Result<ModalResult, String> {
    let dof_num = DofNumbering::build_2d(input);
    let nf = dof_num.n_free;
    let n = dof_num.n_total;

    if nf == 0 {
        return Err("No free DOFs".into());
    }

    let total_mass = compute_total_mass(input, densities);
    if total_mass < 1e-20 {
        return Err("No mass assigned — set material densities".into());
    }

    // Assemble K and M
    let asm = assemble_2d(input, &dof_num);
    let m_full = assemble_mass_matrix_2d(input, &dof_num, densities);

    let free_idx: Vec<usize> = (0..nf).collect();
    let k_ff = extract_submatrix(&asm.k, n, &free_idx, &free_idx);
    let m_ff = extract_submatrix(&m_full, n, &free_idx, &free_idx);

    // Solve K·φ = λ·M·φ where λ = ω²
    let result = lanczos_generalized_eigen(&k_ff, &m_ff, nf, num_modes, 0.0)
        .ok_or_else(|| "Eigenvalue decomposition failed".to_string())?;

    let num_modes = num_modes.min(nf);

    // Build influence vectors for X and Y directions
    // r_x[i] = 1 for X translational DOFs, 0 otherwise
    // r_y[i] = 1 for Y translational DOFs, 0 otherwise
    let mut r_x = vec![0.0; nf];
    let mut r_y = vec![0.0; nf];
    for &node_id in &dof_num.node_order {
        if let Some(&d) = dof_num.map.get(&(node_id, 0)) {
            if d < nf { r_x[d] = 1.0; }
        }
        if let Some(&d) = dof_num.map.get(&(node_id, 1)) {
            if d < nf { r_y[d] = 1.0; }
        }
    }

    let mut modes = Vec::new();
    let mut cum_mrx = 0.0;
    let mut cum_mry = 0.0;

    let n_converged = result.values.len();
    for idx in 0..n_converged {
        let eigenvalue = result.values[idx];
        if eigenvalue <= 1e-10 || modes.len() >= num_modes {
            continue;
        }

        let omega = eigenvalue.sqrt();
        let freq = omega / (2.0 * std::f64::consts::PI);
        let period = if freq > 1e-20 { 1.0 / freq } else { f64::INFINITY };

        // Extract eigenvector (column idx from n×k matrix)
        let phi: Vec<f64> = (0..nf).map(|i| result.vectors[i * n_converged + idx]).collect();

        // Compute φᵀ·M·φ
        let m_phi = mat_vec_sub(&m_ff, &phi, nf);
        let phi_m_phi: f64 = phi.iter().zip(m_phi.iter()).map(|(a, b)| a * b).sum();

        // Participation factors: Γ = φᵀ·M·r / (φᵀ·M·φ)
        let phi_m_rx: f64 = phi.iter().zip(r_x.iter()).zip(m_phi.iter())
            .map(|((_, rx), mp)| {
                // Actually need φᵀ·M·r, but we have M·φ, so use r·(M·φ) ≠ φ·(M·r)
                // Since M is symmetric: φᵀ·M·r = rᵀ·M·φ = Σ r_i * (M·φ)_i
                rx * mp
            }).sum::<f64>();

        let phi_m_ry: f64 = r_y.iter().zip(m_phi.iter())
            .map(|(ry, mp)| ry * mp).sum();

        let gamma_x = if phi_m_phi.abs() > 1e-30 { phi_m_rx / phi_m_phi } else { 0.0 };
        let gamma_y = if phi_m_phi.abs() > 1e-30 { phi_m_ry / phi_m_phi } else { 0.0 };

        // Effective masses
        let meff_x = gamma_x * gamma_x * phi_m_phi;
        let meff_y = gamma_y * gamma_y * phi_m_phi;
        let mrx = if total_mass > 1e-20 { meff_x / total_mass } else { 0.0 };
        let mry = if total_mass > 1e-20 { meff_y / total_mass } else { 0.0 };
        cum_mrx += mrx;
        cum_mry += mry;

        // Build mode shape (normalized to max = 1)
        let mut u_mode = vec![0.0; n];
        let mut max_disp = 0.0f64;
        for i in 0..nf {
            u_mode[i] = phi[i];
            max_disp = max_disp.max(phi[i].abs());
        }
        if max_disp > 1e-20 {
            for val in u_mode.iter_mut().take(nf) {
                *val /= max_disp;
            }
        }

        let displacements = super::linear::build_displacements_2d(&dof_num, &u_mode);

        modes.push(ModeShape {
            frequency: freq,
            period,
            omega,
            displacements,
            participation_x: gamma_x,
            participation_y: gamma_y,
            effective_mass_x: meff_x,
            effective_mass_y: meff_y,
            mass_ratio_x: mrx,
            mass_ratio_y: mry,
        });
    }

    if modes.is_empty() {
        return Err("No valid modes found".into());
    }

    // Rayleigh damping (5% critical from modes 1 and last)
    let rayleigh = if modes.len() >= 2 {
        let w1 = modes[0].omega;
        let w2 = modes.last().unwrap().omega;
        let xi = 0.05; // 5% critical damping
        let a0 = 2.0 * xi * w1 * w2 / (w1 + w2);
        let a1 = 2.0 * xi / (w1 + w2);
        let damping_ratios: Vec<f64> = modes.iter()
            .map(|m| a0 / (2.0 * m.omega) + a1 * m.omega / 2.0)
            .collect();
        Some(RayleighDamping { a0, a1, omega1: w1, omega2: w2, damping_ratios })
    } else {
        None
    };

    Ok(ModalResult {
        modes,
        n_dof: nf,
        total_mass,
        cumulative_mass_ratio_x: cum_mrx,
        cumulative_mass_ratio_y: cum_mry,
        rayleigh,
    })
}

/// M*v product for a submatrix (nf×nf flat array)
fn mat_vec_sub(m: &[f64], v: &[f64], n: usize) -> Vec<f64> {
    let mut result = vec![0.0; n];
    for i in 0..n {
        for j in 0..n {
            result[i] += m[i * n + j] * v[j];
        }
    }
    result
}

File: web/src/lib/engine/modal.ts
// Modal analysis — natural frequencies and mode shapes

import type { SolverInput } from './types';
import type { DofNumbering } from './solver-js';
import { buildDofNumbering, assemble } from './solver-js';
import { assembleMassMatrix } from './mass-matrix';
import { solveGeneralizedEigen, matVec } from './matrix-utils';
import { t } from '../i18n';

export interface ModeShape {
  /** Natural frequency (Hz) */
  frequency: number;
  /** Period (s) */
  period: number;
  /** Angular frequency ω (rad/s) */
  omega: number;
  /** Mode shape displacements per node: { nodeId, ux, uy, rz } normalized */
  displacements: Array<{ nodeId: number; ux: number; uy: number; rz: number }>;
  /** Modal participation factor Γ in X direction */
  participationX: number;
  /** Modal participation factor Γ in Y direction */
  participationY: number;
  /** Effective modal mass in X direction (t = kN·s²/m) */
  effectiveMassX: number;
  /** Effective modal mass in Y direction (t = kN·s²/m) */
  effectiveMassY: number;
  /** Effective mass ratio in X: Meff_x / Mtotal */
  massRatioX: number;
  /** Effective mass ratio in Y: Meff_y / Mtotal */
  massRatioY: number;
}

/** Rayleigh damping coefficients: C = a0·M + a1·K */
export interface RayleighDamping {
  /** Mass-proportional coefficient a₀ */
  a0: number;
  /** Stiffness-proportional coefficient a₁ */
  a1: number;
  /** ω₁ used for computation */
  omega1: number;
  /** ω₂ used for computation */
  omega2: number;
  /** Damping ratios by mode index: ξₙ = a₀/(2ωₙ) + a₁ωₙ/2 */
  dampingRatios: number[];
}

export interface ModalResult {
  modes: ModeShape[];
  /** Number of DOFs in the problem */
  nDof: number;
  /** Total mass of the structure (t = kN·s²/m) */
  totalMass: number;
  /** Cumulative effective mass ratio in X after all computed modes */
  cumulativeMassRatioX: number;
  /** Cumulative effective mass ratio in Y after all computed modes */
  cumulativeMassRatioY: number;
  /** Rayleigh damping for 5% critical (computed from modes 1 & last) */
  rayleigh?: RayleighDamping;
}

/**
 * Solve for natural frequencies and mode shapes.
 * K·φ = ω²·M·φ  →  generalized eigenvalue problem
 *
 * @param input Solver input (structure definition)
 * @param densities Map of materialId → density in kg/m³
 * @param numModes Number of modes to return (default: min(6, nFree))
 */
export function solveModal(
  input: SolverInput,
  densities: Map<number, number>,
  numModes?: number,
): ModalResult | string {
  const dofNum = buildDofNumbering(input);
  const nf = dofNum.nFree;

  if (nf === 0) return t('modal.noFreeDofs');
  if (nf > 500) return t('modal.modelTooLarge');

  // Validate that at least one material has density assigned
  const materialsUsed = new Set<number>();
  for (const elem of input.elements.values()) materialsUsed.add(elem.materialId);
  const missingDensity: number[] = [];
  for (const matId of materialsUsed) {
    const d = densities.get(matId);
    if (d === undefined || d <= 0) missingDensity.push(matId);
  }
  if (missingDensity.length === materialsUsed.size) {
    return t('modal.noDensity');
  }

  // Assemble stiffness (we only need Kff)
  const { K } = assemble(input, dofNum);
  const nt = dofNum.nTotal;
  const Kff = new Float64Array(nf * nf);
  for (let i = 0; i < nf; i++) {
    for (let j = 0; j < nf; j++) {
      Kff[i * nf + j] = K[i * nt + j];
    }
  }

  // Inject densities into a modified input for mass matrix assembly
  // We need materials with density info — create extended solver input
  const inputWithDensity = {
    ...input,
    materials: new Map(
      Array.from(input.materials.entries()).map(([id, mat]) => [
        id,
        { ...mat, density: densities.get(id) ?? 0 },
      ]),
    ),
  };

  const Mff = assembleMassMatrix(inputWithDensity as any, dofNum);

  // Check that mass matrix is not zero
  let massNorm = 0;
  for (let i = 0; i < nf * nf; i++) massNorm += Mff[i] * Mff[i];
  if (massNorm < 1e-20) return t('modal.zeroMassMatrix');

  // Solve generalized eigenvalue problem: Kff·φ = λ·Mff·φ where λ = ω²
  const eigenResult = solveGeneralizedEigen(Kff, Mff, nf);
  if (!eigenResult) return t('modal.choleskyError');

  const nModes = Math.min(numModes ?? 6, nf);
  const modes: ModeShape[] = [];

  // Build influence vectors for participation factors.
  // {r_x} has 1 for every x-translational DOF, 0 otherwise
  // {r_y} has 1 for every y-translational DOF, 0 otherwise
  const rX = new Float64Array(nf);
  const rY = new Float64Array(nf);
  for (const [nodeId] of input.nodes) {
    const ixKey = `${nodeId}:0`;
    const iyKey = `${nodeId}:1`;
    const ixIdx = dofNum.map.get(ixKey);
    const iyIdx = dofNum.map.get(iyKey);
    if (ixIdx !== undefined && ixIdx < nf) rX[ixIdx] = 1;
    if (iyIdx !== undefined && iyIdx < nf) rY[iyIdx] = 1;
  }

  // Total mass: computed from mass matrix diagonal (translational DOFs only).
  // For 2D frames, mass from X-translational DOFs = mass from Y-translational DOFs
  // when all DOFs are free. But when some DOFs are restrained, they differ.
  // Per Chopra §13.2: total mass = trace of M along translational DOFs in one direction.
  // We use the max of X and Y to handle asymmetric restraints.
  const MrX = matVec(Mff, rX, nf);
  const MrY = matVec(Mff, rY, nf);
  let totalMassX = 0;
  for (let i = 0; i < nf; i++) totalMassX += rX[i] * MrX[i];
  let totalMassY = 0;
  for (let i = 0; i < nf; i++) totalMassY += rY[i] * MrY[i];
  // Use max: the direction with fewer restrained DOFs gives the most complete mass
  const totalMassMax = Math.max(totalMassX, totalMassY);

  let cumulativeMeffX = 0;
  let cumulativeMeffY = 0;

  for (let m = 0; m < nModes; m++) {
    const lambda = eigenResult.values[m];
    if (lambda <= 0) continue; // Skip zero/negative eigenvalues

    const omega = Math.sqrt(lambda);
    const frequency = omega / (2 * Math.PI);
    const period = 1 / frequency;

    // Extract RAW mode shape vector (before normalization for display)
    const phiRaw = new Float64Array(nf);
    for (let i = 0; i < nf; i++) {
      phiRaw[i] = eigenResult.vectors[i * nf + m];
    }

    // Compute participation factors BEFORE visual normalization
    // Γₙ_x = φₙᵀ · M · {r_x} / (φₙᵀ · M · φₙ)
    const Mphi = matVec(Mff, phiRaw, nf);
    let phiMphi = 0;
    for (let i = 0; i < nf; i++) phiMphi += phiRaw[i] * Mphi[i];

    let phiMrX = 0;
    for (let i = 0; i < nf; i++) phiMrX += phiRaw[i] * MrX[i];
    let phiMrY = 0;
    for (let i = 0; i < nf; i++) phiMrY += phiRaw[i] * MrY[i];

    const gammaX = phiMphi > 1e-20 ? phiMrX / phiMphi : 0;
    const gammaY = phiMphi > 1e-20 ? phiMrY / phiMphi : 0;

    // Effective modal mass: Meff = Γ² × (φᵀ·M·φ)
    const meffX = gammaX * gammaX * phiMphi;
    const meffY = gammaY * gammaY * phiMphi;

    cumulativeMeffX += meffX;
    cumulativeMeffY += meffY;

    // Normalize for display: max component = 1
    const phi = new Float64Array(phiRaw);
    let maxAbs = 0;
    for (let i = 0; i < nf; i++) {
      if (Math.abs(phi[i]) > maxAbs) maxAbs = Math.abs(phi[i]);
    }
    if (maxAbs > 0) {
      for (let i = 0; i < nf; i++) phi[i] /= maxAbs;
    }

    // Map to node displacements
    const displacements: ModeShape['displacements'] = [];
    for (const [nodeId] of input.nodes) {
      const ux = getPhiComponent(dofNum, phi, nodeId, 0, nf);
      const uy = getPhiComponent(dofNum, phi, nodeId, 1, nf);
      const rz = dofNum.dofsPerNode > 2 ? getPhiComponent(dofNum, phi, nodeId, 2, nf) : 0;
      displacements.push({ nodeId, ux, uy, rz });
    }

    modes.push({
      frequency, period, omega, displacements,
      participationX: gammaX,
      participationY: gammaY,
      effectiveMassX: meffX,
      effectiveMassY: meffY,
      massRatioX: totalMassMax > 1e-20 ? meffX / totalMassMax : 0,
      massRatioY: totalMassMax > 1e-20 ? meffY / totalMassMax : 0,
    });
  }

  if (modes.length === 0) return t('modal.noModesFound');

  // Compute Rayleigh damping coefficients (Chopra §11.4)
  // C = a₀·M + a₁·K, with ξ = a₀/(2ω) + a₁·ω/2
  // Given ξ₁ = ξ₂ = ξ (target damping, default 5%):
  //   a₀ = 2ξ·ω₁·ω₂/(ω₁+ω₂)
  //   a₁ = 2ξ/(ω₁+ω₂)
  let rayleigh: RayleighDamping | undefined;
  if (modes.length >= 2) {
    const xi = 0.05; // 5% critical damping
    const omega1 = modes[0].omega;
    // Use last mode or mode with highest cumulative mass significance
    const omega2 = modes[modes.length - 1].omega;
    if (omega1 > 1e-10 && omega2 > omega1 * 1.01) {
      const a0 = 2 * xi * omega1 * omega2 / (omega1 + omega2);
      const a1 = 2 * xi / (omega1 + omega2);
      // Compute actual damping ratio for each mode
      const dampingRatios = modes.map(m => a0 / (2 * m.omega) + a1 * m.omega / 2);
      rayleigh = { a0, a1, omega1, omega2, dampingRatios };
    }
  }

  return {
    modes,
    nDof: nf,
    totalMass: totalMassMax,
    cumulativeMassRatioX: totalMassMax > 1e-20 ? cumulativeMeffX / totalMassMax : 0,
    cumulativeMassRatioY: totalMassMax > 1e-20 ? cumulativeMeffY / totalMassMax : 0,
    rayleigh,
  };
}

function getPhiComponent(
  dofNum: DofNumbering,
  phi: Float64Array,
  nodeId: number,
  localDof: number,
  nFree: number,
): number {
  if (localDof >= dofNum.dofsPerNode) return 0;
  const key = `${nodeId}:${localDof}`;
  const idx = dofNum.map.get(key);
  if (idx === undefined || idx >= nFree) return 0;
  return phi[idx];
}
